preprocessor.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef PREPROCESSOR_H
+#define PREPROCESSOR_H
+
+#include <string>
+#include <vector>
+
+#include "ast.h"
+
+struct DirectiveState
+{
+    Directive command;
+    bool condition;
+
+    DirectiveState(Directive command_, bool condition_)
+    {
+        command = command_;
+        condition = condition_;
+    }
+};
+
+class Preprocessor
+{
+    std::vector<DirectiveState> stack_;
+    const std::vector<std::string> defines_;
+
+  public:
+    Preprocessor(const std::vector<std::string>& defines) : defines_(defines)
+    {
+        if (!stack_.empty())
+            stack_.clear();
+    }
+
+    ~Preprocessor()
+    {
+        stack_.clear();
+    }
+
+    bool process(Directive cmd, const std::string& arg = "")
+    {
+        bool result = false;
+
+        switch (cmd)
+        {
+            case Ifdef:
+            {
+                DirectiveState state(cmd, false);
+                if (is_defined(arg))
+                    state.condition = true;
+
+                stack_.push_back(state);
+                result = true;
+                break;
+            }
+            case Ifndef:
+            {
+                DirectiveState state(cmd, false);
+                if (!is_defined(arg))
+                    state.condition = true;
+
+                stack_.push_back(state);
+                result = true;
+                break;
+            }
+            case Else:
+            {
+                DirectiveState& current_state = stack_.back();
+                /* Verify that the current state is ifdef or ifndef. */
+                if (current_state.command != Ifdef &&
+                    current_state.command != Ifndef)
+                    break;
+
+                current_state.command = cmd;
+                current_state.condition = !current_state.condition;
+                result = true;
+                break;
+            }
+            case Endif:
+            {
+                /* Ensure the stack is not empty. */
+                if (stack_.empty())
```

Comment:
Should this report an error?

Reply:
I make this function return a boolean value such that the `parser.cpp` can report the error (and provide the line of code).